[Woo POS] Consider an order that is already paid as success in `CollectOrderPaymentUseCase`

jaclync opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #13128 
Based on https://github.com/woocommerce/woocommerce-ios/pull/13165
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

As `capture_terminal_payment` could succeed on the server side but fail returning the response, and if order sync still fails after the graceful handling in https://github.com/woocommerce/woocommerce-ios/pull/13165, the app/POS results in a server error state. After retrying by canceling and collecting payment again, it currently results in an "already paid" error state.

server-side error | after retrying - already paid error
-- | --
![IMG_0390](https://github.com/woocommerce/woocommerce-ios/assets/1945542/1ca1bdee-082f-42a2-b3ef-7b6ce6deddc5) | ![IMG_0391](https://github.com/woocommerce/woocommerce-ios/assets/1945542/647be0a5-959c-4217-8bfa-cdecf9a1e1e6)

If we know that the order is already paid, transitioning to the success state would be a better UX instead of a non-retryable error state in POS. The changes in this PR should only affect POS and not the existing app, since the order might not always be up to date.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

Adapted from https://github.com/woocommerce/woocommerce-ios/pull/13165:

### Existing IPP experience (no changes are expected)

1. Set up Proxyman or Charles to intercept traffic for your test device
2. Launch the app using a store eligible for IPP
3. Navigate to `Orders > +`
4. Add a new order, and tap `Collect Payment`
5. Select `Card reader` and connect to the reader if needed
6. Enable breakpoints in your proxy for the response of `POST` requests to `https://public-api.wordpress.com/rest/v1.1/jetpack-blogs/*/rest-api/` and `GET` requests to `https://public-api.wordpress.com/rest/v1.1/jetpack-blogs/*/rest-api/?json=true&path=/wc/v3/orders/*`
7. Tap your card on the reader
8. Abort the `capture_terminal_payment` response when the breakpoint is hit (equivalent to poor network conditions meaning it was lost.)
9. Shortly after, abort the `orders/*` response when the breakpoint is hit --> the server-side error state should be shown
10. Dismiss the error alert
11. Tap `Collect Payment` for the same order again --> the "already paid" error alert should be shown

<img src="https://github.com/woocommerce/woocommerce-ios/assets/1945542/3ecf8b6b-56cb-4af0-89d6-7cee2d6a7215" width="600" />


### New POS experience

1. Set up Proxyman or Charles to intercept traffic for your test device
2. Launch the app using a store eligible for IPP
3. Navigate to `Menu > Point of Sale mode`
4. Add an item to your basket and tap `Checkout`
5. Enable breakpoints in your proxy for the response of `POST` requests to `https://public-api.wordpress.com/rest/v1.1/jetpack-blogs/*/rest-api/` and `GET` requests to `https://public-api.wordpress.com/rest/v1.1/jetpack-blogs/*/rest-api/?json=true&path=/wc/v3/orders/*`
6. Shortly after, abort the `orders/*` response when the breakpoint is hit --> the server-side error state should be shown
7. Tap `Cancel payment` and then `Collect payment` again
8. Tap your card on the reader
10. Observe that the payment shows as successful

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

After this PR, the success state should be shown:

<img src="https://github.com/woocommerce/woocommerce-ios/assets/1945542/b37705d4-952c-4141-8cc5-28e781b22a25" width="600" />


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

jaclync committed:
If the order is already paid after a remote sync, complete with success.

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Collect Payments/CollectOrderPaymentUseCase.swift:
```diff
@@ -267,7 +267,8 @@ private extension CollectOrderPaymentUseCase {
 
             guard self.isOrderAwaitingPayment() else {
-                return onCheckCompletion(.failure(CollectOrderPaymentUseCaseError.orderAlreadyPaid))
+                return onPaymentCompletion(.success(CardPresentCapturedPaymentData(paymentMethod: .unknown,
```
I don't think we should return success here, in the existing flows. The error is clearer for some of the situations which can come up.

For POS, where we're really confident that the order's just been created, we know that the most likely (perhaps only possible) reason for hitting this error is after retry after the missed network connection as covered in #13165. In that case, the shopper will have just tapped their card, so it's not confusing to see a success payment.

For the existing flow, where we rely more on storage and orders can be much older, it's not so safe to show success here. A merchant could have an outdated copy of the order from storage, tap `Collect Payment`, then immediately see that the order's complete without the customer tapping their card, nor any explanation of how it's complete.

In that situation, they're likely to think that the app is wrong and that they need to redo the order. The error would have explained it much more clearly.

The best place to do this for the new experience is probably in `CardPresentPaymentAlertPresenterAdaptor.present`, e.g. with a case like this one:

```
    case .paymentError(error: CollectOrderPaymentUseCaseError.orderAlreadyPaid, _, _),
                .paymentErrorNonRetryable(error: CollectOrderPaymentUseCaseError.orderAlreadyPaid, _):
            paymentEventSubject.send(.show(eventDetails: .paymentSuccess(done: {})))
```

It probably needs to do more than _just_ that, e.g. we'll need to clean up something in the facade, but that's my idea for a starting point... although from a quick try, it does look like it "just works".

jaclync replied:
> I don't think we should return success here, in the existing flows. The error is clearer for some of the situations which can come up.
> 
> For POS, where we're really confident that the order's just been created, we know that the most likely (perhaps only possible) reason for hitting this error is after retry after the missed network connection as covered in #13165. In that case, the shopper will have just tapped their card, so it's not confusing to see a success payment.
> 
> For the existing flow, where we rely more on storage and orders can be much older, it's not so safe to show success here. A merchant could have an outdated copy of the order from storage, tap `Collect Payment`, then immediately see that the order's complete without the customer tapping their card, nor any explanation of how it's complete.
> 
> In that situation, they're likely to think that the app is wrong and that they need to redo the order. The error would have explained it much more clearly.

Thanks a lot for the explanation about why this change shouldn't be applied to the existing flow, it makes sense. @staskus and I had a pair programming session to work on this issue, and your suggested fix works from my testing! (I tested that this scenario results in success for POS and remains an error in the existing flow in order creation).

> It probably needs to do more than just that, e.g. we'll need to clean up something in the facade, but that's my idea for a starting point... although from a quick try, it does look like it "just works".

For this, I'd like to check with you. There are also [2](https://github.com/woocommerce/woocommerce-ios/blob/802b6bb208d62810f11868ab328051a30ab27130/WooCommerce/Classes/POS/Card%20Present%20Payments/CardPresentPaymentsAlertPresenterAdaptor.swift#L24) [other](https://github.com/woocommerce/woocommerce-ios/blob/802b6bb208d62810f11868ab328051a30ab27130/WooCommerce/Classes/POS/Card%20Present%20Payments/CardPresentPaymentsAlertPresenterAdaptor.swift#L33) TODO comments about calling the facade's `cancelPayment` for further cleanup. I tried triggering `CardPresentPaymentService.cancelPayment` in the event's `cancelPayment` closure, but the current implementation of cancellation is to also call the event's `cancelPayment` closure:

https://github.com/woocommerce/woocommerce-ios/blob/802b6bb208d62810f11868ab328051a30ab27130/WooCommerce/Classes/POS/Card%20Present%20Payments/CardPresentPaymentCollectOrderPaymentUseCaseAdaptor.swift#L144-L146

In `CardPresentPaymentCollectOrderPaymentUseCaseAdaptor`, we could also change how we handle the cancellation. An option is to cancel the payment intent, which I think we could do for non-retryable payment errors (there could be exceptions that I didn't know of). Does this sound like the right way to go, canceling the payment intent for non-retryable payment errors (moving [this line](https://github.com/woocommerce/woocommerce-ios/blob/802b6bb208d62810f11868ab328051a30ab27130/WooCommerce/Classes/POS/Card%20Present%20Payments/CardPresentPaymentCollectOrderPaymentUseCaseAdaptor.swift#L145) to the .processing cases below)?

For retryable payment errors, is there any cleanup we should do? In place of [this TODO comment](https://github.com/woocommerce/woocommerce-ios/blob/802b6bb208d62810f11868ab328051a30ab27130/WooCommerce/Classes/POS/Card%20Present%20Payments/CardPresentPaymentsAlertPresenterAdaptor.swift#L24).

If this is a bit too much to take on before your AFK, I can create a separate task for this further cancelation cleanup for later.

joshheald replied:
For this case I don't think it would be appropriate or needed to call `cancelPayment`. The `orderAlreadyPaid` error is thrown before we create the payment intent, so there's nothing to cancel yet.

From our testing, it's working without further cleanup, so that's great. It makes sense actually, because we throw away the `CollectOrderPaymentUseCase` when the merchant taps `New transaction`.

However, `CollectOrderPaymentUseCase` doesn't know anything about this `success`, so we're tracking this as a payment failure in our analytics... or actually two failures:

![CleanShot 2024-06-28 at 12 06 37@2x](https://github.com/woocommerce/woocommerce-ios/assets/2472348/dac8415a-86c8-4079-9550-18d6d3da0e15)

We're not really dealing with this yet, but it would be nice to track success if we _can_. This can be delayed but let's capture it in a ticket. It's fine to track the failures, but we should track the success as well at the end.

I think perhaps we should defer those TODOs for after my AFK – we'll have a new project to tighten everything up and deal with the TODOs... I can't dig in to them now.

jaclync replied:
> However, `CollectOrderPaymentUseCase` doesn't know anything about this `success`, so we're tracking this as a payment failure in our analytics... or actually two failures:
> 
> ![CleanShot 2024-06-28 at 12 06 37@2x](https://private-user-images.githubusercontent.com/2472348/344140191-dac8415a-86c8-4079-9550-18d6d3da0e15.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk1ODEyNjksIm5iZiI6MTcxOTU4MDk2OSwicGF0aCI6Ii8yNDcyMzQ4LzM0NDE0MDE5MS1kYWM4NDE1YS04NmM4LTQwNzktOTU1MC0xOGQ2ZDNkYTBlMTUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYyOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MjhUMTMyMjQ5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjUwYmJiZjRkZTIxMmNlZDMzNWRjMzMzNGNiNmIxZDVhNTQxZDM5NjNkYjNiOTQ3NTdkNjAzMzZkMDAzZWUzYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.U7I-cFD-ZhKJP21y4tEpCBeukvKIQht5BjJCi_sJGVg)
> 
> We're not really dealing with this yet, but it would be nice to track success if we _can_. This can be delayed but let's capture it in a ticket. It's fine to track the failures, but we should track the success as well at the end.

Good catch on the tracking side, yea we'd want to make sure to track the success state for this case. I created a [draft task](https://github.com/orgs/woocommerce/projects/247/views/6?sliceBy%5Bvalue%5D=iOS+%F0%9F%8D%8E&pane=issue&itemId=69143334) for this, currently in the "Blocked" column because we don't have an analytics plan yet. I suggested tracking POS success when the view model receives the payment success event and is about to show the success UI.

joshheald reviewed (commented):
I don't think this is the right approach, because of the change in the existing app.

I think there, we should still show an "order already paid" error, and adapt it in the POS only.

More detail in the in-line comment.

staskus committed:
Revert "If the order is already paid after a remote sync, complete with success."

This reverts commit 1b883708d65ca8c3561d23ade391e836ee51720f.

staskus committed:
Handle orderAlreadyPaid error separately for POS use case

jaclync committed:
Move special handling on specific retryable/non-retryable errors before the default error handling because Swift goes evaluates switch cases in order.

jaclync commented:
Thanks for working on the changes and pushing the commits @staskus, I added a commit to move the handling of the "already paid" error before the default error handling because Swift executes switch cases in order.

@joshheald I've updated the PR & description if you get a chance to review it! If not, feel free to just briefly share what you think makes the most sense - leaving the PR open, creating a separate issue for further cleanup, etc.

jaclync commented:
Thanks for reviewing & testing this Josh 🙏  merging it now!

jaclync committed:
Merge branch 'trunk' into feat/13128-handle-already-paid

* trunk: (97 commits)
  Update naming for analytics card to display a stat and list of top performers
  Add generic stats data formatting helper to format amounts with currency
  Bump version number
  Fix rubocop violation
  Refine error wording
  Allow `finalize_release` lane to be retried on CI
  Bump version number
  Update metadata translations
  Update app translations – `Localizable.strings`
  Update tests for GoogleAdsStore
  Export GoogleAdsCampaign from Yosemite
  Update Google ads action and store to support fetching campaigns
  Fix swiftlint issue
  Update MockGoogleListingsAndAdsRemote
  Regenerate fake and copiable
  Add support for requesting paginated stats data
  Add support for decoding paginated campaign stats
  Support stats responses with a subset of totals fields
  Always send fields and orderby params with stats request
  feat: use reverse-DNS keys in localized strings
  ...
